R/QFeatures-class.R

```diff
@@ -667,36 +690,268 @@ longFormat <- function(object,
 }
 
 
-##' @param name A `character(1)` naming the single assay (default is
-##'     `"newAssay"`). Ignored if `y` is a list of assays.
+##' @param y An object that inherits from `SummarizedExperiment` or a
+##'     *named* list of assays. When `y` is a list, each element must
+##'     inherit from a `SummarizedExperiment` and the names of the 
+##'     list are used as the names of the assays to add. Hence, the 
+##'     list names must be unique and cannot overlap with the names of
+##'     the assays already present in `x`.
+##' 
+##' @param name A `character(1)` naming the single assay. Ignored if 
+##'     `y` is a list of assays.
+##'     
 ##' @param assayLinks An optional [AssayLinks].
-##'
+##' 
 ##' @md
 ##'
 ##' @rdname QFeatures-class
 ##'
 ##' @export
 addAssay <- function(x,
                      y,
-                     name = "newAssay",
-                     assayLinks = AssayLinks(names = name)) {
+                     name,
+                     assayLinks) {
+    ## Check arguments
+    stopifnot(inherits(x, "QFeatures"))
+    y <- .checkAssaysToInsert(y, x, name, replace = FALSE)
+    
+    ## Check (or create) assayLinks
+    if (!missing(assayLinks)) {
+        if (inherits(assayLinks, "AssayLink"))
+            assayLinks <- AssayLinks(assayLinks)
+        if (!identical(sort(names(assayLinks)), sort(names(y))))
+            stop("'assayLinks' must be named after the assay(s) in 'y'.")
+    } else {
+        assayLinks <- AssayLinks(names = names(y))
+    }
+    
+    ## Update the colData
+    cd <- .updateColData(x, y)
+    
+    ## Add the assay to the ExperimentList
+    ## NOTE: we replace using the `@` slot. Although not recommended, 
+    ## this bypasses the checks of all the elements (using 
+    ## `validObject`) in the ExperimentList as this is already 
+    ## performed when building the QFeatures object and `y` is checked
+    ## at the beginning of the function. This leads to a reduction in
+    ## computational time. 
+    el <- experiments(x)
+    for(ii in names(y)) {
+        el@listData[[ii]] <- y[[ii]]
```

Comment:
Should we consider checking the validity of the new assays?

Reply:
Yes, this is performed at line 716

Reply:
Could you add a link to `.checkAssaysToInsert()`.

Reply:
Sorry here it is:  https://github.com/cvanderaa/QFeatures/blob/2e314fb5c32ed393d46506affbaf7ed0b0d4b244/R/QFeatures-class.R#L865-L893